ts/Series/Treegraph/TreegraphLayout.ts

```diff
@@ -0,0 +1,367 @@
+                        dummyNode.shift = 0;
+                        dummyNode.thread = void 0 as any;
+                        dummyNode.change = 0;
+                        dummyNode.preX = 0;
+
+                        // add connection from new node to the previous points
+
+                        // first connection to itself:
+                        dummyNode.linksFrom.push(merge(parent.linksFrom[0]));
+                        dummyNode.linksFrom[0].fromNode = dummyNode;
+                        dummyNode.linksTo.push(merge(parent.linksFrom[0]));
+                        dummyNode.linksTo[0].toNode = dummyNode;
+                        dummyNode.linksTo[0].fromNode = parent;
+                        dummyNode.linksFrom[0].toNode = child;
+                        // temporary:
+                        dummyNode.linksTo[0].to = void 0 as any;
+
+                        dummyNode.column = child.column - gapSize;
+                        dummyNode.level = child.level - gapSize;
+
+                        // Then connection from parent to dummyNode
+                        if (!parent.linksFrom[index].trueToNode) {
+                            parent.linksFrom[index].trueToNode = child;
+                        }
+                        parent.linksFrom[index].toNode = dummyNode;
+
+                        // Then connection from chhild to dummyNode
+                        if (!child.linksTo[0].trueFromNode) {
+                            child.linksTo[0].trueFromNode = parent;
+                        }
+                        child.linksTo[0].fromNode = dummyNode;
+                        return dummyNode;
+                    };
+                    while (gapSize > 0) {
+                        child = createDummyNode(parent, child);
+                        gapSize--;
+                    }
+                }
+            });
+        });
+    },
+    // Init the values of the node. probably redundant, since this can be
+    // performed in the init of the chart, but might be useful for repositioning
+    // of the nodes on the update.
+    initValues: function (nodes: TreegraphNode[]): void {
+        nodes.forEach((node: TreegraphNode): void => {
+            node.mod = 0;
+            node.ancestor = node;
+            node.shift = 0;
+            node.thread = void 0 as any;
+            node.change = 0;
+            node.preX = 0;
+        });
+    },
+    // Adds the value to each node, which indicates, what is his sibling number.
+    calculateInitialX: function (
+        this: Highcharts.TreegraphLayout,
+        node: TreegraphNode,
+        index: number
+    ): void {
+        const treeLayout = this as Highcharts.TreegraphLayout;
+        node.linksFrom.forEach((link, index): void => {
+            treeLayout.calculateInitialX(link.toNode, index);
+        });
+        node.relativeXPosition = index;
+    },
+    // recursive post order traversal of the tree, where the initial position of
+    // the node is calculated.
```

Comment:
Do we want the layout algorithm to be following our API docs structure? I am wondering if that should be visible to our customers. What do you think?

Reply:
it could, but none of these methods would make sense on their own. I guess, that all methods should be made private apart from `calculatePositions`

Reply:
nvm, you can't do it, if you are extending the algorithm the way we are doing it.